proposed/swift-interop.md

```diff
@@ -0,0 +1,183 @@
+
+##### Tuples
+
+If possible, Swift tuples should be represented as `ValueTuple`s in .NET. If this is not possible, then they should be represented as types with a `Deconstruct` method similar to `ValueTuple` to allow a tuple-like experience in C#.
+
+#### Projection Tooling Components
+
+The projection tooling should be split into these components:
+
+##### Importing Swift into .NET
+
+1. A tool that takes in a `.swiftinterface` file or Swift sources and produces C# code.
+2. A library that provides the basic support for Swift interop that the generated code builds on.
+3. User tooling to easily generate Swift projections for a given set of `.framework`s.
+    - This tooling would build a higher-level interface on top of the tool in item 1 that is more user-friendly and project-system-integrated.
+4. (optional) A NuGet package, possibly referencable by `FrameworkReference` or automatically included when targeting macOS, Mac Catalyst, iOS, or tvOS platforms that exposes the platform APIs for each `.framework` that is exposed from Swift to .NET.
+    - This would be required to provide a single source of truth for Swift types so they can be exposed across an assembly boundary.
+
+#####  Exporting .NET to Swift
+
+There are two components to exporting .NET to Swift: Implementing existing Swift types in .NET and passing instances of those types to Swift, and exposing novel types from .NET code to Swift code to be created from Swift. Exposing novel types from .NET code to Swift code is considered out of scope at this time.
+
+For implementing existing Swift types in .NET, we will require one of the following tooling options:
+
+1. A Roslyn source generator to generate any supporting code needed to produce any required metadata, such as type metadata and witness tables, to pass instances of Swift-type-implementing .NET types defined in the current project to Swift.
+2. An IL-post-processing tool to generate the supporting code and metadata from the compiled assembly.
+
+An IL-post-processing tool would not integrate well with Hot-Reload support, so a Roslyn source-generator is likely a better option.
```

Comment:
> integrate well with Hot-Reload support

This is an assumption I would like to confirm. Is there latitude with how Swift operates where users can reload types that Swift may already know about?

Reply:
What I am getting at, is HotReload a tenable feature for autogenerated interop code? Where are the boundaries?

Reply:
Swift has a Hot Reload feature for SwiftUI, but Swift in general doesn't look to have a general Hot Reload feature. Changing the contents of a method body (a la Edit-and-Continue style edits) on the .NET side should be something we can make work (especially as .NET on iOS and .NET on macOS currently supports Hot Reload), but adding members that make a type adopt new protocols likely wouldn't work well.

The biggest problem with using a post-processing tool is not that it would break Hot-Reload for the Swift interop code, but that it would break Hot-Reload for the assembly as a whole, including in .NET-only portions of the code, due to the IL post-processing step.

Reply:
> but adding members that make a type adopt new protocols likely wouldn't work well.

Or fields.

> The biggest problem with using a post-processing tool is not that it would break Hot-Reload for the Swift interop code, but that it would break Hot-Reload for the assembly as a whole, including in .NET-only portions of the code, due to the IL post-processing step.

Yep, that is the clarity we need here.